File: src/main/java/com/demo/t_web/program/user/model/UserRole.java
package com.demo.t_web.program.user.model;

import com.fasterxml.jackson.annotation.JsonBackReference;
import jakarta.persistence.*;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;
import org.springframework.security.core.GrantedAuthority;

import java.io.Serializable;
import java.util.Date;

/**
 * <pre>
 * com.demo.t_web.program.login.model.UserRoles
 *   - UserRoles.java
 * </pre>
 *
 * @author : tarr4h
 * @className : UserRoles
 * @description :
 * @date : 12/13/24
 */
@Entity
@Table(name = "adp_user_roles")
@Getter
@Setter
@NoArgsConstructor
public class UserRole implements GrantedAuthority {

    @EmbeddedId
    private RoleId id;

    @Column(name = "role_name")
    private String roleName;

    @Column(name = "reg_dt")
    @Temporal(TemporalType.TIMESTAMP)
    private Date regDt;

    @ManyToOne
    @MapsId("userId")
    @JoinColumn(name = "user_id")
    @JsonBackReference
    private User user;

    @PrePersist
    public void prePersist(){
        if(regDt == null){
            regDt = new Date();
        }
    }

    @Override
    public String getAuthority() {
        return getId().getRoleId();
    }

    @Embeddable
    @Getter
    @Setter
    @NoArgsConstructor
    public static class RoleId implements Serializable {

        public RoleId(String roleId, String userId) {
            this.roleId = roleId;
            this.userId = userId;
        }

        @Column(name = "role_id")
        private String roleId;
        @Column(name = "user_id")
        private String userId;
    }

    public void setId(String roleId, String userId){
        this.id = new RoleId(roleId, userId);
    }

    public UserRole(User user, String roleId, String roleName){
        setId(roleId, user.getId());
        this.roleName = roleName;
        this.user = user;
    }
}

File: src/main/java/com/demo/t_web/program/user/model/UserRole.kt
package com.demo.t_web.program.user.model

import com.demo.t_web.program.sys.model.BaseVo
import com.fasterxml.jackson.annotation.JsonBackReference
import jakarta.persistence.*
import org.springframework.security.core.GrantedAuthority
import java.io.Serializable

@Entity
@Table(name = "adp_user_roles")
data class UserRole(

    @EmbeddedId
    var id : RoleId? = null,

    @Column(name = "role_name")
    var roleName: String = "",

    @ManyToOne
    @MapsId("userId")
    @JoinColumn(name = "user_id")
    @JsonBackReference
    var user: User? = null,
) : BaseVo(), GrantedAuthority {

    constructor(user: User, roleId: String, roleName: String) : this() {
        setId(roleId, user.id)
        this.roleName = roleName
        this.user = user
    }

    override fun getAuthority() = id?.roleId

    companion object {
        @Embeddable
        data class RoleId(
            @Column(name = "role_id")
            var roleId : String = "",
            @Column(name = "user_id")
            var userId : String = "",
        ) : Serializable
    }

    fun setId(roleId: String, userId: String){
        this.id = RoleId(roleId, userId)
    }
}
